### PR TITLE
Fix throwing exception when calling RunClassConstructor on a generic type with a static constructor

### LINKED ISSUE #103891: System.InvalidOperationException: Could not execute the method because either the method itself or the containing type is not fully instantiated.
### Description

Calling RuntimeHelpers.RunClassConstructor on a generic type throws an exception.

### Reproduction Steps

Test case:

```cs
using System.Runtime.CompilerServices;

class App
{
    static void Main()
    {
        foreach (var type in typeof (App).Assembly.GetTypes ())
            RuntimeHelpers.RunClassConstructor (type.TypeHandle);
    }

    static App ()
    {
        Console.WriteLine ("App cctor!");
    }
}

class C<T> {
    static C () {
        Console.WriteLine ("cctor!");
    }
}
```

### Expected behavior

```shell
$ dotnet run
App cctor!
```

### Actual behavior

```shell
dotnet run
App cctor!
Unhandled exception. System.InvalidOperationException: Could not execute the method because either the method itself or the containing type is not fully instantiated.
   at C`1..cctor()
   at System.Runtime.CompilerServices.RuntimeHelpers.RunClassConstructor(RuntimeTypeHandle type)
   at App.Main() in /Users/rolf/test/dotnet/console/Program.cs:line 8
```

### Regression?

Yes, this works in .NET 8 and 9.0.0-preview.5.24256.1, but fails with 9.0.100-preview.6.24320.12

It started happening in this maestro bump; https://github.com/xamarin/xamarin-macios/pull/20621, so the regression range is: https://github.com/dotnet/runtime/compare/84b3339505...3750ac5161

### Known Workarounds

_No response_

### Configuration

_No response_

### Other information

_No response_

### PR DESCRIPTION
**Pull request by steveisok**
https://github.com/dotnet/runtime/pull/99183 seems to have done away with assuming that a generic type's static constructor was always "initialized". As a result, if you call RunClassConstructor on it, the runtime would throw an exception.

Fixes https://github.com/dotnet/runtime/issues/103891

### CODE REVIEW

File: src/coreclr/vm/methodtable.cpp [hunk 1]
```diff
@@ -3768,6 +3768,10 @@ void MethodTable::CheckRunClassInitThrowing()
     if (IsSharedByGenericInstantiations())
         return;
 
+    // For back-compat reasons, act like it has already been initialized
```
Comment (jkotas) [inline src/coreclr/vm/methodtable.cpp:3771]: I do not think we need the check here. It can be assert.

**Comment by jkotas** [comment]
This needs a test.

E.g. add `RuntimeHelpers.RunClassConstructor(typeof(GenericHasCctor<>).TypeHandle);` to https://github.com/dotnet/runtime/blob/7134e1e6a48dcddf1bc3578765c9471c7f91d2c3/src/libraries/System.Runtime/tests/System.Runtime.Tests/System/Runtime/CompilerServices/RuntimeHelpersTests.cs#L104

**Review by lambdageek — APPROVED** [review]

File: src/coreclr/vm/methodtable.cpp [hunk 2]
```diff
@@ -3768,6 +3768,9 @@ void MethodTable::CheckRunClassInitThrowing()
...
```
Comment (jkotas) [inline src/coreclr/vm/methodtable.cpp:3771]:
```suggestion
```

File: src/libraries/System.Runtime/tests/System.Runtime.Tests/System/Runtime/CompilerServices/RuntimeHelpersTests.cs [hunk 3]
```diff
@@ -101,6 +101,8 @@ public static void RunClassConstructor()
             RuntimeTypeHandle t = typeof(HasCctor).TypeHandle;
             RuntimeHelpers.RunClassConstructor(t);
             Assert.Equal("Hello", HasCctorReceiver.S);
+            // Should not throw
+            RuntimeHelpers.RunClassConstructor(typeof(GenericHasCctor<>).TypeHandle);
             return;
```
Comment (jkotas) [inline src/libraries/System.Runtime/tests/System.Runtime.Tests/System/Runtime/CompilerServices/RuntimeHelpersTests.cs:106]:
```suggestion
```

File: src/libraries/System.Runtime/tests/System.Runtime.Tests/System/Runtime/CompilerServices/RuntimeHelpersTests.cs [hunk 4]
```diff
@@ -101,7 +101,8 @@ public static void RunClassConstructor()
             RuntimeTypeHandle t = typeof(HasCctor).TypeHandle;
             RuntimeHelpers.RunClassConstructor(t);
             Assert.Equal("Hello", HasCctorReceiver.S);
-            return;
+            // Should not throw
+            RuntimeHelpers.RunClassConstructor(typeof(GenericHasCctor<>).TypeHandle);
```
Comment (jkotas) [inline src/libraries/System.Runtime/tests/System.Runtime.Tests/System/Runtime/CompilerServices/RuntimeHelpersTests.cs:105]: We also need something like this below:
```
internal class GenericHasCctor<T>
{
    static GenericHasCctor()
    {
        Thread.Yield(); // Make sure the preinitialization optimization doesn't eat this.
    }
}

**Review by jkotas — APPROVED** [review]

**Pull request merged**